news/_posts/2019-09-11-prestashop-in-2019-and-beyond-part-2-pain-points.md

```diff
@@ -0,0 +1,471 @@
+
+We have already started improving this in PrestaShop 1.7.6, but there is still some way to go. Having two translation systems working in parallel (the legacy one and the Symfony one) is very confusing.
+
+## Testing is very hard
+
+![Testing is doubting][testing-is-doubting]{: style="height:400px"}
+_Testing is doubting, by [CommitStrip](http://www.commitstrip.com/)_
+{: .text-center }
+
+This is one of the classic _consequences_ of complex, static, globally-stateful code. Since components are so tightly interweaved and depend on global state, putting the system in a state that allows to test a feature in an isolated and predictable way is really, really hard. In some cases, dependencies and hidden state can become extremely difficult and time-consuming to understand, specially when dealing with static in-memory cache (ie. variables that maintain their value between calls to a same method on a same thread).
+
+Unsurprisingly, hard-to-test code also leads to complex, hard-to-maintain test scripts.
+
+Ideally, we should be able to classify tests into three types:
+
+- **Unit** – Where a single, completely isolated class is tested at a time.
+- **Integration** – Where a subsystem is tested in its entirety, with carefully selected dependencies replaced by test doubles.
+- **E2E** – Where the system is tested in its entirety from a user standpoint, using browser automation.
+
+Here's what we currently have:
+
+- A mix of PhpUnit-based, unit/integration/survival tests, located in the `tests-legacy` folder (bound to be rewritten).
+- PhpUnit-based Unit tests, located in `tests/Unit`.
+- PhpUnit-based Integration tests, located in `tests/Integration`.
+- Behat-based Behavior tests, located in `tests/Integration/Behaviour`.
+- Two stacks of End-to-end (E2E) tests, based on Mocha and Selenium:
+	- An old one called "starter theme", located in the `tests/Selenium` folder that we are phasing out.
+	- A more recent one in the `tests/E2E` folder.
+
+We have been struggling with automated tests failing unexpectedly, not because the feature they tested had stopped working, but because of diverse, unrelated technical issues:
+
+- Some unit tests aren't really unit (usually because they tested several classes at once).
+- Many unit tests that aren't unit were also extremely hard to understand and maintain, because of all the boilerplate code needed to put the shop in the state we needed for testing.
+- Some tests were producing side-effects that made other tests fail because of unexpected/inconsistent state (usually due to hidden cache in legacy code).
+- E2E tests fail randomly because of timeouts and/or issues with our task runner (Travis).
+
+We believe that bugs are a natural side-effect of changing things, and that automated tests are the only way to be able to consistently move forward without breaking stuff... but for that, we need good-quality tests.
```

Comment:
```suggestion
We believe that bugs are a natural side-effect of changing things, and that automated tests are the only way to be able to consistently move forward without breaking stuff... but for that, we need good-quality tests. Because of PrestaShop legacy architecture, these tests are really hard to write.
```